sdk/src/main/scala/com/horizen/account/forger/AccountForgeMessageBuilder.scala

```diff
@@ -140,9 +141,91 @@ class AccountForgeMessageBuilder(mainchainSynchronizer: MainchainSynchronizer,
     ommersSerializer.toBytes(ommers.asJava).length
   }
 
-  override def getForgingStakeMerklePathInfo(nextConsensusEpochNumber: ConsensusEpochNumber, wallet:  AccountWallet): Seq[ForgingStakeMerklePathInfo] =
-    ???
+  // TODO the stateRoot of the genesis block
+  def getGenesisBlockRootHash =
+    new Array[Byte](32)
 
+  def getStateRoot(history: AccountHistory, nextConsensusEpochNumber: ConsensusEpochNumber) : (ConsensusEpochNumber, Array[Byte]) = {
+
+    // For given epoch N we should get data from the ending of the epoch N-2.
+    // genesis block is the single and the last block of epoch 1 - that is a special case:
+    // Data from epoch 1 is also valid for epoch 2, so for epoch N==2, we should get info from epoch 1.
+    nextConsensusEpochNumber match {
+      case epoch if (epoch <= 2) =>
+        (ConsensusEpochNumber @@ 1, getGenesisBlockRootHash)
+
+      case epoch =>
+
+        val branchPointInfo: BranchPointInfo = getBranchPointInfo(history) match {
+          case Success(info) => info
+          case Failure(ex) =>
+            throw ex
+        }
+
+        // apply two times the api for getting previous epoch last block
+        val parentId_0 = branchPointInfo.branchPointId
+        val ts_0 = history.blockInfoById(parentId_0).timestamp
+        val blockId_0 = history.getLastBlockInPreviousConsensusEpoch(ts_0, parentId_0)
+
+        val parentId_1 = blockId_0
+        val ts_1 = history.blockInfoById(parentId_1).timestamp
+        val blockId_1 = history.getLastBlockInPreviousConsensusEpoch(ts_1, parentId_1)
+
+        val lastBlock = history.getBlockById(blockId_1)
+        val stateRoot = lastBlock.get().header.stateRoot
+
+        (ConsensusEpochNumber @@ (epoch - 2), stateRoot)
+    }
+  }
+
+  override def getForgingStakeMerklePathInfo(
+         nextConsensusEpochNumber: ConsensusEpochNumber,
+         wallet: AccountWallet,
+         history: AccountHistory,
+         state: AccountState): Seq[ForgingStakeMerklePathInfo] = {
+
+    // 1. get from history the state root from the header of the block of 2 epochs before
+    val (consensusEpochNumber, stateRoot) = getStateRoot(history, nextConsensusEpochNumber)
+
+    // 2. get from stateDb using root above the collection of all forger stakes (ordered)
+    val stateViewFromRoot = if (ByteUtils.toHexString(stateRoot) == ByteUtils.toHexString(getGenesisBlockRootHash)) {
+      // in case of genesis block
+      state.getView
+    } else {
+      state.getStateDbViewFromRoot(stateRoot)
+    }
+
+    val forgingStakeInfoSeq : Seq[ForgingStakeInfo] = stateViewFromRoot.getOrderedForgingStakeInfoSeq()
+
+    // 3. using wallet secrets, filter out the not-mine forging stakes
+     val secrets : Seq[Secret] = wallet.allSecrets().asScala
+
+    val walletPubKeys = secrets.map( e => e.publicImage())
+
+    val filteredForgingStakeInfoSeq = forgingStakeInfoSeq.filter(p => walletPubKeys.contains(p.blockSignPublicKey))
+
+    val epochInfo = ConsensusEpochInfo(
```

Comment:
don't need the whole instance, only the tree `MerkleTree.createMerkleTree(fil....`

Reply:
Ok, done